contents/security/cryptography.md

```diff
@@ -0,0 +1,81 @@
+
+One of the earliest encryption algorithms or ciphers is the substitution cipher, which encrypts text by substituting each letter of the message with another letter. This page introduces the [Caesar cipher](http://www.cs.trincoll.edu/~crypto/historical/caesar.html), a substitution cipher, and how it can be defeated using statistical analysis. 
+Modern ciphers today operate on bits rather than letters.
+
+## Symmetric Key Ciphers
+
+In a symmetric key cipher, the same key, a shared secret, is used for both encryption and decryption.
+The parties wishing to communicate securely share the same key. For example, in the substitution cipher, the same key (the substitution table) is used for both encryption and decryption.
+
+### Stream ciphers
+
+The ciphertext, keystream and  plaintext are sequences or streams of bits of equal length. 
+During encryption, each plaintext bit is combined (usually xored) with the corresponding keystream bit to produce the ciphertext bits. For example, the 1st plaintext bit is xored with the 1st keystream bit, 2nd plaintext bit with 2nd keystream bit and so on.
+During decryption, each ciphertext bit is xored with the corresponding keystream bit to produce the plaintext.
+Notice that encryption and decryption are the same operation; this is possible since xoring a bit x with another bit y twice recovers the original bit; (x xor y) xor y = x.
+
+The simplest stream cipher is the [one-time pad](https://en.wikipedia.org/wiki/One-time_pad). In this cipher, the keystream used is bits from a truly random source, and is also the key.
+It is the only known cipher that cannot be cracked, even if the attacker has infinite computing power.
+This property is known as [perfect secrecy](https://crypto.stackexchange.com/questions/3896/simply-put-what-does-perfect-secrecy-mean); the ciphertext gives no additional information about the plaintext, so knowing the ciphertext does not provide any advantage to the attacker trying to recover the plaintext.
+Unfortunately, Shannon proved that any cipher that achieves perfect secrecy has the following limitations, making them [impractical](https://www.schneier.com/crypto-gram/archives/2002/1015.html#7).
+
+* The key must be truely random, not pseudorandomly generated, and must never be reused.
+* The key must be securely distributed, and be at least as long as the message being generated. For example, to send a 10gb file to someone encrypted with the one-time pad requires sending 10gb of key material + 10gb of ciphertext = 20gb.
+
+In practice, we do not require perfect secrecy, since attackers have limited computational power. Hence, all other ciphers are only secure against such attackers.
```

Comment:
try rephrasing "Hence, all other ciphers are only secure against such attackers." it's a bit confusing